design/src/rfcs/rfc0028_serialization_and_deserialization.md

```diff
@@ -0,0 +1,349 @@
+We conclude that we must avoid such a technical challenge to bring this RFC to life.
+
+## `fn set_fields` to allow users supply externally created `Input`s
+
+SDK does not have a method that allows users supply deserialized inputs.
+Thus, we add a new method `fn set_fields` to `Client` types.
+This method accepts inputs and replace all parameters that `Client` has with the new one.
+
+```rust
+pub fn set_fields(mut self, builder: path::to::builder_type) -> path::to::builder_type {
+    self.inner = new_parameter;
+    self
+}
+```
+
+# Other Concerns
+## Model evolution
+SDK will introduce new fields and it is possible that we may see new data types in future.
+
+We believe that this will not be a problem.
+
+### Introduction of New Fields
+Most fields are `Option<T>` type.
+When user de-serializes data written for a format before the new fields were introduced, new field will be assigned with `None` type.
+
+If a field isn't `Option`, `serde` uses `Default` trait unless a custom de-serialization/serialization is specified to generate data to fill the field.
+If the new field is not an `Option<T>` type and has no `Default` implementation, we must implement a custom de-serialization logic.
+
+In case of serilization, introduction of new fields will not be an issue unless the data format requires a schema. (e.g. parquet, avro) However, this is outside the scope of this RFC.
+
+## Introduction of New Data Type
+If a new field introduces a new data type, it will not require any additional work if the data type can derive `serde` traits.
+
+If the data cannot derive `serde` traits on it's own, then we have two options.
+To clarify, this is the same approach we took on `Data Type to skip` section.
+1. skip
+   We will simply skip serializing/de-serializing. However, we may need to implement custom serialization/de-serialization logic if a value is not wrapped with `Option`.
+2. custom serialization/de-serialization logic
+   We can implement tailored serialization/de-serialization logic.
+
+Either way, we will mention this on the generated docs to avoid suprising users.
+
+e.g.
+```rust
+#[derive(serde::Serialize, serde::Deserialize)]
+struct OutputV1 {
+  string_field: Option<String>
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct OutputV2 {
+  string_field: Option<String>,
+  // this will always be treated as None value by serde
+  #[serde(skip)]
+  skip_not_serializable: Option<SomeComplexDataType>,
+  // We can implement a custom serialization logic
+  #[serde(serialize_with = "custom_serilization_logic", deserialize_with = "custom_deserilization_logic")]
+  not_derive_able: SomeComplexDataType,
+  // Serialization will be skipped, and de-serialization will be handled with the function provided on default tag
+  #[serde(skip, default = "default_value")]
+  skip_with_custom: DataTypeWithoutDefaultTrait,
+}
+```
+# Discussions
+
+## Serialization and de-serialization support for an entire response/request
```

Comment:
I added this section in response to @rcoh 's comment 
https://github.com/awslabs/smithy-rs/pull/1944#pullrequestreview-1217927496

> One last major stumbling block in this design—this does not solve the problem of response serialization/deserialization. Omitting this is acceptable, but if we do so, we need to clearly state why in the RFC. 

I wanted to clarify that this RFC does not aim to provide a way to serialize response
Just to make it absolutely clear, I wanted to say that I'm not going to do that for requests as well.

let me fix up language.

Reply:
@rcoh - What exactly did you mean by this?

Reply:
I think he meant things like headers, and other http related stuff.

Reply:
My guess is that he meant the wire/protocol serialized format for actual requests/responses to/from AWS is distinct from this serde format, and that we have no intention of making them consistent.

Reply:
@rcoh 
I appreciate clarification.

Reply:
by utilizing fluent builders, you enable customers to use serde to serialize _requests_ (e.g. `dynamodb::PutItem`)

But this design doesn't provide an easy way to extend it to responses/outputs (because the fluent builders here only match the request shape, not the output shape, e.g. there is no easy way to serialize the _response_ you would get from `dynamodb::ScanItems`

Reply:
I'm not sure if I understand you correctly.

Are you saying that we can't serialize data types that fluent builder's `send` function returns?

Like `ScanOutput` found here.
```rust
pub async fn send(
    self,
) -> std::result::Result<
    crate::output::ScanOutput,
    aws_smithy_http::result::SdkError<crate::error::ScanError>,
> {
    let op = self
        .inner
        .build()
        .map_err(aws_smithy_http::result::SdkError::construction_failure)?
        .make_operation(&self.handle.conf)
        .await
        .map_err(aws_smithy_http::result::SdkError::construction_failure)?;
    self.handle.client.call(op).await
}
```

As I stated on *## Builder Types and Non-Builder Types*, this RFC aims to implement serde traits onto Output shapes as well.

dynamodb's scan returns `ScanOutput` (I'm assuming that this is what you mean by dynamodb::ScanItems.), and this data type is going to be implemented with `serde` traits as well.

I looked up the `ScanOutput` but I don't think there will be an issue deriving serde for it.

https://docs.rs/aws-sdk-dynamodb/0.22.0/src/aws_sdk_dynamodb/output.rs.html#859-884

Reply:
ah I didn't realize that deriving serde traits on outputs was in the plan—I think that's OK, I'll need to think on it

Reply:
gotcha. I think I will just delete this section.

Reply:
fixed